presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java

```diff
@@ -34,7 +34,18 @@ default boolean isEnabled(Session session)
         return true;
     }
 
-    Optional<PlanNode> apply(T node, Captures captures, Context context);
+    @Deprecated
+    default Optional<PlanNode> apply(T node, Captures captures, Context context)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    // TODO remove default implementaiton once deprecated apply method got removed
+    default void apply(T node, Captures captures, Context context, Result result)
```

Comment:
> For instance, think of JoinEnumeration. It could return enumerations on the go and when it returns enumeration for which cost is acceptable, then the optimizer could break such computations.

This would be only possible if (a) you use interrupts and trust them to be effective (I don't) or (b) you return an iterator (or equiv.) which optimizer would be free to stop consuming at some point.

> [...] They make the code harder to reason about, harder to understand concurrency semantics

+1